### PR TITLE
Demo for amp-web-push component

### PR DESCRIPTION
**Pull request by demianrenzulli**
AMP by Example demo for [amp-web-push](https://github.com/ampproject/amphtml/blob/master/extensions/amp-web-push/amp-web-push.md) component.

- This demo requires two extra static files: a [helper iframe](https://cdn.ampproject.org/v0/amp-web-push-helper-frame.html) and a [permission dialog](https://cdn.ampproject.org/v0/amp-web-push-permission-dialog.html) (placed at the /static/webpush/ folder).
- It follows the [VAPID standard](https://blog.mozilla.org/services/2016/04/04/using-vapid-with-webpush/) for Webpush.
- Uses an IndexedDB database ("web-push-db"), to store and retrieve the [Webpush subscription](https://developer.mozilla.org/en-US/docs/Web/API/PushSubscription) object, to avoid storing it on a remote DB.
- The SW had to be updated to include [the events](https://github.com/ampproject/amphtml/blob/master/extensions/amp-web-push/0.1/amp-web-push.service-worker.js) suggested in the documentation, to store and retrieve the tokens and listen to "push" events.
- The SW will also intercept the request from the form of the page for the "send-push" event, retrieve the token from IndexedDB and send a request to an endpoint at: https://demo-amp-web-push.glitch.me/send-push, so the notification can be created and sent in the backend with the [WebPush Node library](https://github.com/web-push-libs/web-push).

### CODE REVIEW

File: src/20_Components/amp-web-push.html [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    <!--
+      Clicking the subscription widget pops up a page prompting the user for notification permissions and signals the service worker (configured below) to subscribe the user to push in the background. On this example, we are intercepting the call to `onMessageReceivedSubscribe`, and storing the subscription object on IndexedDB.
+    -->
+    <amp-web-push-widget visibility="unsubscribed" layout="fixed" width="500" height="70">
+        <button class="ampstart-btn m1" on="tap:amp-web-push.subscribe">Subscribe to Notifications</button>
+    </amp-web-push-widget>
+
+    <!--
+      Clicking the unsubscription widget signals the worker to unsubscribe the user from push in the background. On this example, we'll be also remvoing the previoulsy stored subscription object from IndexedDB.
```
Comment (kul3r4) [inline src/20_Components/amp-web-push.html]: remvoing -> removing

File: static/sw.js [hunk 2]
```diff
@@ -15,6 +15,7 @@
  **/
 
 importScripts('/sw-toolbox/sw-toolbox.js');
+importScripts("https://cdn.rawgit.com/jakearchibald/idb/97e4e878/lib/idb.js");
```
Comment (kul3r4) [inline static/sw.js]: can you use single quotes?

File: src/20_Components/amp-web-push.html [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+    <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript>
+        <style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style>
+    </noscript>
+    <!-- ## Setup -->
+    <!--
+      Import the `amp-web-push` component in the header.
+    -->
+    <script async custom-element="amp-web-push" src="https://cdn.ampproject.org/v0/amp-web-push-0.1.js"></script>
+    <script async custom-element="amp-form" src="https://cdn.ampproject.org/v0/amp-form-0.1.js"></script>
```
Comment (kul3r4) [inline src/20_Components/amp-web-push.html]: could you add a comment for each import explaining why you need it?

**Comment by demianrenzulli** [comment]
Thanks for your comments, I'll work on them.

I deployed a version here: https://amp-web-push-demo.firebaseapp.com/components/amp-web-push/.

On that version, playground is not working, as I only deployed the static part.
Please, let me know if you need anything else.

File: src/20_Components/amp-web-push.html [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+<body>
+    <!-- ## Basic Usage -->
+    <!--
+    The amp-web-push component requires two additional HTML files as well as an amp-web-push compatible Service Worker. These three files form the configuration described below.
+    -->
+    <amp-web-push id="amp-web-push" layout="nodisplay" helper-iframe-url="<%host%>/webpush/amp-web-push-helper-frame.html" permission-dialog-url="<%host%>/webpush/amp-web-push-permission-dialog.html" service-worker-url="<%host%>/sw.js">
+    </amp-web-push>
+
+    <!--
+      Clicking the subscription widget pops up a page prompting the user for notification permissions and signals the service worker (configured below) to subscribe the user to push in the background. On this example, we are intercepting the call to `onMessageReceivedSubscribe`, and storing the subscription object on IndexedDB.
```
Comment (kul3r4) [inline src/20_Components/amp-web-push.html]: check the rendered version on https://amp-web-push-demo.firebaseapp.com/components/amp-web-push/, the doc is not rendered correctly. Moving "Clicking..." to a row above should fix it.

File: src/20_Components/amp-web-push.html (hunk 1 above)
Comment (kul3r4) [reply src/20_Components/amp-web-push.html]: same as above for docs

File: src/20_Components/amp-web-push.html [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    <script async custom-element="amp-form" src="https://cdn.ampproject.org/v0/amp-form-0.1.js"></script>
+    <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
+    <script async custom-element="amp-install-serviceworker" src="https://cdn.ampproject.org/v0/amp-install-serviceworker-0.1.js"></script>
+    <script async src="https://cdn.ampproject.org/v0.js"></script>
+</head>
+
+<body>
+    <!-- ## Basic Usage -->
+    <!--
+    The amp-web-push component requires two additional HTML files as well as an amp-web-push compatible Service Worker. These three files form the configuration described below.
```
Comment (kul3r4) [inline src/20_Components/amp-web-push.html]: amp-web-push -> `amp-web-push`

**Comment by kul3r4** [comment]
Thanks Demian for the preview, I left some other comments.

**Comment by demianrenzulli** [comment]
This is great, thanks Chiara. I'll include your recommendations on my update.

**Comment by demianrenzulli** [comment]
Hi @kul3r4 just made changes based on your comments and uploaded a new version here:
https://amp-web-push-demo.firebaseapp.com/components/amp-web-push/

**Comment by kul3r4** [comment]
Looks good

File: src/20_Components/amp-web-push.html [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    <amp-web-push-widget visibility="subscribed" layout="fixed" width="500" height="180">
+        <button class="ampstart-btn m1" on="tap:amp-web-push.unsubscribe">Unsubscribe from Notifications</button>
+        <form method="post" class="m1" action-xhr="/send-push" target="_top">
+            <input type="submit" value="Send Web Push" class="ampstart-btn">
+            <br/><br/>
+            <div submitting>
+                Sending Push message...
+            </div>
+            <div submit-success>
+                <template type="amp-mustache">
```
Comment (sebastianbenz) [inline src/20_Components/amp-web-push.html]: You can remove the template here

File: src/20_Components/amp-web-push.html [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                    Web Push sent!
+                </template>
+            </div>
+            <div submit-error>
+                <template type="amp-mustache">
```
Comment (sebastianbenz) [inline src/20_Components/amp-web-push.html]: You can remove the template here

File: src/20_Components/amp-web-push.html [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    -->
+    <script async custom-element="amp-web-push" src="https://cdn.ampproject.org/v0/amp-web-push-0.1.js"></script>
+    <!--
+      The `amp-form` component is being used in this demo, to send a post request so the notification can be prepared and sent.
+    -->
+    <script async custom-element="amp-form" src="https://cdn.ampproject.org/v0/amp-form-0.1.js"></script>
+    <!--
+      The `amp-mustache` component is used in this demo to show success and error messages after received response for the form submission.
+    -->
+    <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
```
Comment (sebastianbenz) [inline src/20_Components/amp-web-push.html]: You don't need amp-mustache

File: src/20_Components/amp-web-push.html [hunk 9]
```diff
@@ -0,0 +1,69 @@
+<!---{
+  "disablePlayground": true
```
Comment (sebastianbenz) [inline src/20_Components/amp-web-push.html]: I'm not happy with disabling the playground as it's the only way to view the whole source code on the page. I'd prefer leaving it in. I can look into why it's not working.

**Comment by sebastianbenz** [comment]
Can you re-enable the playground. Otherwise looks great! Thanks a lot for your work!

**Comment by demianrenzulli** [comment]
@sebastianbenz thank you for the help! I just enabled Playground, please, let me know if I can do something else.

**Comment by sebastianbenz** [comment]
👍

**Pull request merged**